source/neuropod/bindings/c/c_api.cc

```diff
@@ -18,32 +18,126 @@ limitations under the License.
+                             NP_Status *              status)
 {
-    *model = new NP_Neuropod();
+    try
+    {
+        *model          = new NP_Neuropod();
+        (*model)->model = std::make_unique<neuropod::Neuropod>(neuropod_path);
+        if (status != nullptr)
+        {
+            NP_ClearStatus(status);
+        }
+    }
+    catch (std::exception &e)
+    {
+        if (status != nullptr)
+        {
+            status->code    = NEUROPOD_ERROR;
+            status->message = e.what();
+        }
+    }
+}
+
+NP_RuntimeOptions NP_DefaultRuntimeOptions()
+{
+    // Use C++ runtime options object to set default values.
+    neuropod::RuntimeOptions default_options;
+    NP_RuntimeOptions        options;
+
+    options.use_ope                         = default_options.use_ope;
+    options.visible_device                  = static_cast<NP_RuntimeOptions::NP_Device>(default_options.visible_device);
+    options.load_model_at_construction      = default_options.load_model_at_construction;
+    options.disable_shape_and_type_checking = default_options.disable_shape_and_type_checking;
+
+    auto ope_options                     = &options.ope_options;
+    ope_options->free_memory_every_cycle = default_options.ope_options.free_memory_every_cycle;
+    // Control queue name is empty, a new worker will be started.
+    // This is what is expected by default.
+    ope_options->control_queue_name[0] = '\0';
 
-    (*model)->model = std::make_unique<neuropod::Neuropod>(neuropod_path);
+    return options;
+}
+
+void NP_LoadNeuropod(const char *neuropod_path, NP_Neuropod **model, NP_Status *status)
+{
+    const auto &options = NP_DefaultRuntimeOptions();
+    NP_LoadNeuropodWithOpts(neuropod_path, &options, model, status);
 }
 
-// Free a model
 void NP_FreeNeuropod(NP_Neuropod *model)
 {
     delete model;
 }
 
-// Run inference
-// Note: The caller is responsible for freeing the returned NP_NeuropodValueMap
 void NP_Infer(NP_Neuropod *model, const NP_NeuropodValueMap *inputs, NP_NeuropodValueMap **outputs, NP_Status *status)
 {
-    *outputs         = new NP_NeuropodValueMap();
-    (*outputs)->data = std::move(*model->model->infer(inputs->data));
+    NP_InferWithRequestedOutputs(model, inputs, 0, nullptr, outputs, status);
+}
+
+void NP_InferWithRequestedOutputs(NP_Neuropod *              model,
+                                  const NP_NeuropodValueMap *inputs,
+                                  size_t                     noutputs,
+                                  const char **              requested_outputs,
+                                  NP_NeuropodValueMap **     outputs,
+                                  NP_Status *                status)
+{
+    try
+    {
+        // By default empty collection of requested_ouputs expected.
+        std::vector<std::string> rout;
+        for (size_t i = 0; i < noutputs; ++i)
+        {
+            rout.push_back(requested_outputs[i]);
```

Comment:
Also infer found an issue (see the failing lint check):

```

neuropod/bindings/c/c_api.cc:86: error: NULL_DEREFERENCE
--
  | pointer `null` is dereferenced by call to `NP_InferWithRequestedOutputs()` at line 86, column 5.
  | 84.   void NP_Infer(NP_Neuropod *model, const NP_NeuropodValueMap *inputs, NP_NeuropodValueMap **outputs, NP_Status *status)
  | 85.   {
  | 86. >     NP_InferWithRequestedOutputs(model, inputs, 0, nullptr, outputs, status);
  | 87.   }
  | 88.
  |  
  |  
  | Summary of the reports
  |  
  | NULL_DEREFERENCE: 1

```

I know in the example it gave, `noutputs` is 0, but we should probably add a null check on `requested_outputs` anyway

Reply:
added checks